wgpu-core/src/binding_model.rs

```diff
@@ -446,25 +446,64 @@ pub type BindGroupLayouts<A> = crate::storage::Storage<BindGroupLayout<A>, BindG
 ///  - produced bind groups
 ///  - produced pipeline layouts
 ///  - pipelines with implicit layouts
-#[derive(Debug)]
 pub struct BindGroupLayout<A: hal::Api> {
-    pub(crate) raw: A::BindGroupLayout,
     pub(crate) device_id: Stored<DeviceId>,
     pub(crate) multi_ref_count: MultiRefCount,
-    pub(crate) entries: BindEntryMap,
     // When a layout created and there already exists a compatible layout the new layout
     // keeps a reference to the older compatible one. In some places we substitute the
     // bind group layout id with its compatible sibling.
     // Since this substitution can come at a cost, it is skipped when wgpu-core generates
     // its own resource IDs.
-    pub(crate) compatible_layout: Option<Valid<BindGroupLayoutId>>,
+    pub(crate) inner: BglOrDuplicate<A>,
+}
+
+pub(crate) enum BglOrDuplicate<A: hal::Api> {
+    Inner(BindGroupLayoutInner<A>),
+    Duplicate(Valid<BindGroupLayoutId>),
+}
+
+pub struct BindGroupLayoutInner<A: hal::Api> {
+    pub(crate) raw: A::BindGroupLayout,
+    pub(crate) entries: BindEntryMap,
     #[allow(unused)]
     pub(crate) dynamic_count: usize,
     pub(crate) count_validator: BindingTypeMaxCountValidator,
     #[cfg(debug_assertions)]
     pub(crate) label: String,
 }
 
+impl<A: hal::Api> BindGroupLayout<A> {
+    pub(crate) fn assume_deduplicated(&self) -> &BindGroupLayoutInner<A> {
+        match self.inner {
+            BglOrDuplicate::Inner(ref inner) => inner,
+            BglOrDuplicate::Duplicate(_) => {
+                panic!()
```

Comment:
It's rather that the function is self-explanatory enough that we don't need a string to paraphrase what the problem is about.
Calling `assume_deduplicated` better expresses what invariant we have upheld for things to work (this will never panic if we did the de-duplication beforehand), than calling `as_inner().unwrap()` which only says "I'm confident this contains inner bind group layout data".